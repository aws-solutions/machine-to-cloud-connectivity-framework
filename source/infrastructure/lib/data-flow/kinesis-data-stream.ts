// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { KinesisStreamsToKinesisFirehoseToS3 } from '@aws-solutions-constructs/aws-kinesisstreams-kinesisfirehose-s3';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { NagSuppressions } from 'cdk-nag';
import { Construct } from 'constructs';

export interface KinesisDataStreamConstructProps {
  readonly s3LoggingBucket: Bucket;
}

/**
 * Creates a Kinesis data stream, a Kinesis data firehose, a data S3 bucket, and related roles.
 */
export class KinesisDataStreamConstruct extends Construct {
  public kinesisStreamName: string;
  public dataBucketName: string;

  constructor(scope: Construct, id: string, props: KinesisDataStreamConstructProps) {
    super(scope, id);

    const kinesisStreamsToKinesisFirehoseToS3 = new KinesisStreamsToKinesisFirehoseToS3(this, 'DataStream', {
      existingLoggingBucketObj: props.s3LoggingBucket,
      bucketProps: {
        serverAccessLogsPrefix: 'm2c2data/'
      }
    });
    this.kinesisStreamName = kinesisStreamsToKinesisFirehoseToS3.kinesisStream.streamName;
    this.dataBucketName = (<Bucket>kinesisStreamsToKinesisFirehoseToS3.s3Bucket).bucketName;

    // cdk-nag suppressions
    NagSuppressions.addResourceSuppressions(
      kinesisStreamsToKinesisFirehoseToS3,
      [
        { id: 'AwsSolutions-IAM5', reason: 'It specifies a bucket, so it is not a wildcard permission.' },
        { id: 'AwsSolutions-KDF1', reason: 'The data stream is encrypted.' }
      ],
      true
    );
  }
}
