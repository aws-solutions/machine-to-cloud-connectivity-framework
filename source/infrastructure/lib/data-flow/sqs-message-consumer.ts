// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { IotToSqs, IotToSqsProps } from '@aws-solutions-constructs/aws-iot-sqs';
import { LambdaToDynamoDB } from '@aws-solutions-constructs/aws-lambda-dynamodb';
import { SqsToLambda } from '@aws-solutions-constructs/aws-sqs-lambda';
import { CfnDeletionPolicy, Duration } from 'aws-cdk-lib';
import { AttributeType, CfnTable, Table } from 'aws-cdk-lib/aws-dynamodb';
import { Code, Runtime } from 'aws-cdk-lib/aws-lambda';
import { IBucket } from 'aws-cdk-lib/aws-s3';
import { NagSuppressions } from 'cdk-nag';
import { Construct } from 'constructs';

export interface SQSMessageConsumerConstructProps {
  readonly solutionConfig: {
    loggingLevel: string;
    solutionId: string;
    solutionVersion: string;
    sourceCodeBucket: IBucket;
    sourceCodePrefix: string;
  };
}

/**
 * Creates a message SQS queue, a message consumer Lambda function, a logs DynamoDB tables, and IoT rules.
 */
export class SQSMessageConsumerConstruct extends Construct {
  public logsTable: Table;
  private readonly sql: string = `SELECT topic(3) as connectionName, topic(2) as logType, timestamp() as timestamp, * FROM 'm2c2/{type}/+'`;

  constructor(scope: Construct, id: string, props: SQSMessageConsumerConstructProps) {
    super(scope, id);

    const sourceCodeBucket = props.solutionConfig.sourceCodeBucket;
    const sourceCodePrefix = props.solutionConfig.sourceCodePrefix;

    const sqsToLambda = new SqsToLambda(this, 'SQSMessageConsumer', {
      lambdaFunctionProps: {
        code: Code.fromBucket(sourceCodeBucket, `${sourceCodePrefix}/sqs-message-consumer.zip`),
        description: 'Machine to Cloud Connectivity SQS message consumer function',
        environment: {
          LOGGING_LEVEL: props.solutionConfig.loggingLevel,
          SOLUTION_ID: props.solutionConfig.solutionId,
          SOLUTION_VERSION: props.solutionConfig.solutionVersion
        },
        handler: 'sqs-message-consumer/index.handler',
        retryAttempts: 0,
        runtime: Runtime.NODEJS_14_X,
        timeout: Duration.minutes(1)
      },
      queueProps: {
        visibilityTimeout: Duration.minutes(1)
      },
      maxReceiveCount: 3
    });

    const lambdaToDynamoDb = new LambdaToDynamoDB(this, 'LogsDynamoDB', {
      existingLambdaObj: sqsToLambda.lambdaFunction,
      dynamoTableProps: {
        partitionKey: {
          name: 'connectionName',
          type: AttributeType.STRING
        },
        sortKey: {
          name: 'timestamp',
          type: AttributeType.NUMBER
        },
        timeToLiveAttribute: 'ttl'
      },
      tableEnvironmentVariableName: 'LOGS_DYNAMODB_TABLE'
    });
    this.logsTable = lambdaToDynamoDb.dynamoTable;

    const cfnDynamoTable = <CfnTable>this.logsTable.node.defaultChild;
    cfnDynamoTable.cfnOptions.deletionPolicy = CfnDeletionPolicy.DELETE;

    const infoLogsProps: IotToSqsProps = {
      existingQueueObj: sqsToLambda.sqsQueue,
      iotTopicRuleProps: {
        topicRulePayload: {
          actions: [],
          awsIotSqlVersion: '2016-03-23',
          description: 'Processing connection info logs',
          sql: this.sql.replace('{type}', 'info')
        }
      }
    };
    new IotToSqs(this, 'InfoLogsRule', infoLogsProps);

    const errorLogsProps: IotToSqsProps = {
      existingQueueObj: sqsToLambda.sqsQueue,
      iotTopicRuleProps: {
        topicRulePayload: {
          actions: [],
          awsIotSqlVersion: '2016-03-23',
          description: 'Processing connection error logs',
          sql: this.sql.replace('{type}', 'error')
        }
      }
    };
    new IotToSqs(this, 'ErrorLogsRule', errorLogsProps);

    // cdk-nag suppressions
    NagSuppressions.addResourceSuppressions(
      sqsToLambda,
      [
        { id: 'AwsSolutions-IAM5', reason: 'It does not allow wildcard permissions' },
        { id: 'AwsSolutions-SQS3', reason: 'deadLetterQueue is the dead letter queue.' }
      ],
      true
    );
  }
}
