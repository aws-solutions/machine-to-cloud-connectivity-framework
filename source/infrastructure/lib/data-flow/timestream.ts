// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { KinesisStreamsToLambda } from '@aws-solutions-constructs/aws-kinesisstreams-lambda';
import { CfnCondition, CfnDeletionPolicy, Duration, Fn, CustomResource, CfnCustomResource, Aws } from 'aws-cdk-lib';
import { Effect, Policy, PolicyStatement, Role } from 'aws-cdk-lib/aws-iam';
import { Code, Runtime } from 'aws-cdk-lib/aws-lambda';
import { CfnDatabase, CfnTable } from 'aws-cdk-lib/aws-timestream';
import { IBucket } from 'aws-cdk-lib/aws-s3';
import { NagSuppressions } from 'cdk-nag';
import { Construct } from 'constructs';
import { addCfnSuppressRules } from '../../utils/utils';

export interface TimestreamConstructProps {
  readonly existingDatabaseName: string;
  readonly solutionConfig: {
    loggingLevel: string;
    solutionId: string;
    solutionVersion: string;
    sourceCodeBucket: IBucket;
    sourceCodePrefix: string;
    uuid: string;
  };
  readonly customResourcesFunctionArn: string;
  readonly shouldTeardownData: CfnCondition;
}

/**
 * Creates a Lambda function, Kinesis Data Stream streams, Timestream database and table, and other related resources.
 */
export class TimestreamConstruct extends Construct {
  public kinesisStreamArn: string;
  public kinesisStreamName: string;
  public timestreamDatabaseTable: string;

  constructor(scope: Construct, id: string, props: TimestreamConstructProps) {
    super(scope, id);

    const sourceCodeBucket = props.solutionConfig.sourceCodeBucket;
    const sourceCodePrefix = props.solutionConfig.sourceCodePrefix;

    const createTimestreamDatabaseCondition = new CfnCondition(this, 'CreateTimestreamDatabase', {
      expression: Fn.conditionEquals(props.existingDatabaseName, '')
    });

    const timestreamDatabase = new CfnDatabase(this, 'Database', {
      databaseName: `${Aws.STACK_NAME}-${props.solutionConfig.uuid}`
    });
    timestreamDatabase.cfnOptions.condition = createTimestreamDatabaseCondition;
    timestreamDatabase.cfnOptions.deletionPolicy = CfnDeletionPolicy.RETAIN;

    const timestreamDatabaseName = Fn.conditionIf(
      createTimestreamDatabaseCondition.logicalId,
      timestreamDatabase.ref,
      props.existingDatabaseName
    ).toString();

    const teardownTimestreamDatabase = new CustomResource(this, 'teardownTimestreamDatabase', {
      serviceToken: props.customResourcesFunctionArn,
      properties: {
        Resource: 'DeleteTimestreamDatabase',
        DatabaseName: timestreamDatabaseName
      }
    });
    const cfnTeardownGreengrassResourcesBucket = <CfnCustomResource>teardownTimestreamDatabase.node.defaultChild;
    cfnTeardownGreengrassResourcesBucket.cfnOptions.condition = props.shouldTeardownData;

    const timestreamTable = new CfnTable(this, 'Table', {
      databaseName: timestreamDatabaseName,
      retentionProperties: {
        MemoryStoreRetentionPeriodInHours: 24 * 90,
        MagneticStoreRetentionPeriodInDays: 365
      }
    });
    const timestreamTableName = timestreamTable.getAtt('Name').toString();
    const timestreamTableArn = timestreamTable.getAtt('Arn').toString();
    timestreamTable.cfnOptions.deletionPolicy = CfnDeletionPolicy.RETAIN;

    const kinesisStreamsToLambda = new KinesisStreamsToLambda(this, 'KinesisLambda', {
      lambdaFunctionProps: {
        code: Code.fromBucket(sourceCodeBucket, `${sourceCodePrefix}/timestream-writer.zip`),
        description: 'Machine to Cloud Connectivity Framework Timestream data writer function',
        environment: {
          LOGGING_LEVEL: props.solutionConfig.loggingLevel,
          SOLUTION_ID: props.solutionConfig.solutionId,
          SOLUTION_VERSION: props.solutionConfig.solutionVersion,
          TIMESTREAM_DATABASE: timestreamDatabaseName,
          TIMESTREAM_TABLE: timestreamTableName
        },
        handler: 'timestream-writer/index.handler',
        runtime: Runtime.NODEJS_16_X,
        timeout: Duration.seconds(30)
      }
    });

    const lambdaFunctionPolicy = new Policy(this, 'TimestreamPolicy', {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ['timestream:WriteRecords'],
          resources: [timestreamTableArn]
        }),
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ['timestream:DescribeEndpoints'],
          resources: ['*']
        })
      ]
    });

    addCfnSuppressRules(lambdaFunctionPolicy, [
      { id: 'W12', reason: 'timestream:DescribeEndpoints cannot have specific resources.' }
    ]);

    const lambdaFunctionRole = <Role>kinesisStreamsToLambda.lambdaFunction.role;
    lambdaFunctionRole.attachInlinePolicy(lambdaFunctionPolicy);

    this.kinesisStreamArn = kinesisStreamsToLambda.kinesisStream.streamArn;
    this.kinesisStreamName = kinesisStreamsToLambda.kinesisStream.streamName;
    this.timestreamDatabaseTable = `${timestreamDatabaseName}.${timestreamTableName}`;

    // cdk-nag suppressions
    NagSuppressions.addResourceSuppressions(
      kinesisStreamsToLambda,
      [
        { id: 'AwsSolutions-IAM5', reason: 'It does not allow wildcard permissions.' },
        { id: 'AwsSolutions-SQS3', reason: 'SqsDlqQueue is the dead letter queue.' }
      ],
      true
    );
    NagSuppressions.addResourceSuppressions(timestreamDatabase, [
      { id: 'AwsSolutions-TS3', reason: 'The default KMS is used by default.' }
    ]);
    NagSuppressions.addResourceSuppressions(lambdaFunctionPolicy, [
      { id: 'AwsSolutions-IAM5', reason: 'timestream:DescribeEndpoints cannot have specific resources.' }
    ]);
  }
}
