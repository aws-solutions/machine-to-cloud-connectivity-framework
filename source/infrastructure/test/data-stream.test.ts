// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import '@aws-cdk/assert/jest';
import { SynthUtils } from '@aws-cdk/assert';
import { Stack } from 'aws-cdk-lib';
import { Bucket, CfnBucket } from 'aws-cdk-lib/aws-s3';
import { KinesisDataStreamConstruct } from '../lib/data-flow/kinesis-data-stream';

test('M2C2 data stream test', () => {
  const stack = new Stack();
  const s3LoggingBucket = new Bucket(stack, 'TestLoggingBucket');
  (<CfnBucket>s3LoggingBucket.node.defaultChild).overrideLogicalId('TestLoggingBucket');
  const kinesisDataStream = new KinesisDataStreamConstruct(stack, 'TestKinesis', { s3LoggingBucket });

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  expect(kinesisDataStream.kinesisStreamName).toBeDefined();
  expect(kinesisDataStream.dataBucketName).toBeDefined();
  expect(stack).toHaveResourceLike('AWS::S3::Bucket', {
    LoggingConfiguration: {
      DestinationBucketName: {
        Ref: 'TestLoggingBucket'
      },
      LogFilePrefix: 'm2c2data/'
    }
  });
});
