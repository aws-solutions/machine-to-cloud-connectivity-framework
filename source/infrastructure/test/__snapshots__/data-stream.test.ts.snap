// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`M2C2 data stream test 1`] = `
Object {
  "Resources": Object {
    "TestKinesisDataStreamKinesisFirehoseToS3KinesisFirehose65C9C06A": Object {
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-IAM5",
              "reason": "It specifies a bucket, so it is not a wildcard permission.",
            },
            Object {
              "id": "AwsSolutions-KDF1",
              "reason": "The data stream is encrypted.",
            },
          ],
        },
      },
      "Properties": Object {
        "DeliveryStreamType": "KinesisStreamAsSource",
        "ExtendedS3DestinationConfiguration": Object {
          "BucketARN": Object {
            "Fn::GetAtt": Array [
              "TestKinesisDataStreamKinesisFirehoseToS3S3BucketBAAD035E",
              "Arn",
            ],
          },
          "BufferingHints": Object {
            "IntervalInSeconds": 300,
            "SizeInMBs": 5,
          },
          "CloudWatchLoggingOptions": Object {
            "Enabled": true,
            "LogGroupName": Object {
              "Ref": "TestKinesisDataStreamKinesisFirehoseToS3firehoseloggroup044C9C02",
            },
            "LogStreamName": Object {
              "Ref": "TestKinesisDataStreamKinesisFirehoseToS3firehoseloggroupfirehoselogstream1435D54F",
            },
          },
          "CompressionFormat": "GZIP",
          "EncryptionConfiguration": Object {
            "KMSEncryptionConfig": Object {
              "AWSKMSKeyARN": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":kms:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":alias/aws/s3",
                  ],
                ],
              },
            },
          },
          "RoleARN": Object {
            "Fn::GetAtt": Array [
              "TestKinesisDataStreamKinesisFirehoseToS3KinesisFirehoseRoleB8514A3C",
              "Arn",
            ],
          },
        },
        "KinesisStreamSourceConfiguration": Object {
          "KinesisStreamARN": Object {
            "Fn::GetAtt": Array [
              "TestKinesisDataStreamKinesisStream0CE70927",
              "Arn",
            ],
          },
          "RoleARN": Object {
            "Fn::GetAtt": Array [
              "TestKinesisKinesisStreamsRoleDDF9DCCB",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::KinesisFirehose::DeliveryStream",
    },
    "TestKinesisDataStreamKinesisFirehoseToS3KinesisFirehosePolicy965E5D12": Object {
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-IAM5",
              "reason": "It specifies a bucket, so it is not a wildcard permission.",
            },
            Object {
              "id": "AwsSolutions-KDF1",
              "reason": "The data stream is encrypted.",
            },
          ],
        },
      },
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:AbortMultipartUpload",
                "s3:GetBucketLocation",
                "s3:GetObject",
                "s3:ListBucket",
                "s3:ListBucketMultipartUploads",
                "s3:PutObject",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "TestKinesisDataStreamKinesisFirehoseToS3S3BucketBAAD035E",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "TestKinesisDataStreamKinesisFirehoseToS3S3BucketBAAD035E",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "logs:PutLogEvents",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":logs:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":log-group:",
                    Object {
                      "Ref": "TestKinesisDataStreamKinesisFirehoseToS3firehoseloggroup044C9C02",
                    },
                    ":log-stream:",
                    Object {
                      "Ref": "TestKinesisDataStreamKinesisFirehoseToS3firehoseloggroupfirehoselogstream1435D54F",
                    },
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestKinesisDataStreamKinesisFirehoseToS3KinesisFirehosePolicy965E5D12",
        "Roles": Array [
          Object {
            "Ref": "TestKinesisDataStreamKinesisFirehoseToS3KinesisFirehoseRoleB8514A3C",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestKinesisDataStreamKinesisFirehoseToS3KinesisFirehoseRoleB8514A3C": Object {
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-IAM5",
              "reason": "It specifies a bucket, so it is not a wildcard permission.",
            },
            Object {
              "id": "AwsSolutions-KDF1",
              "reason": "The data stream is encrypted.",
            },
          ],
        },
      },
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "firehose.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "TestKinesisDataStreamKinesisFirehoseToS3S3BucketBAAD035E": Object {
      "DeletionPolicy": "Retain",
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-IAM5",
              "reason": "It specifies a bucket, so it is not a wildcard permission.",
            },
            Object {
              "id": "AwsSolutions-KDF1",
              "reason": "The data stream is encrypted.",
            },
          ],
        },
      },
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "LifecycleConfiguration": Object {
          "Rules": Array [
            Object {
              "NoncurrentVersionTransitions": Array [
                Object {
                  "StorageClass": "GLACIER",
                  "TransitionInDays": 90,
                },
              ],
              "Status": "Enabled",
            },
          ],
        },
        "LoggingConfiguration": Object {
          "DestinationBucketName": Object {
            "Ref": "TestLoggingBucket",
          },
          "LogFilePrefix": "m2c2data/",
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "VersioningConfiguration": Object {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "TestKinesisDataStreamKinesisFirehoseToS3S3BucketPolicy54723E0B": Object {
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-IAM5",
              "reason": "It specifies a bucket, so it is not a wildcard permission.",
            },
            Object {
              "id": "AwsSolutions-KDF1",
              "reason": "The data stream is encrypted.",
            },
          ],
        },
      },
      "Properties": Object {
        "Bucket": Object {
          "Ref": "TestKinesisDataStreamKinesisFirehoseToS3S3BucketBAAD035E",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:*",
              "Condition": Object {
                "Bool": Object {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "TestKinesisDataStreamKinesisFirehoseToS3S3BucketBAAD035E",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "TestKinesisDataStreamKinesisFirehoseToS3S3BucketBAAD035E",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "TestKinesisDataStreamKinesisFirehoseToS3firehoseloggroup044C9C02": Object {
      "DeletionPolicy": "Retain",
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-IAM5",
              "reason": "It specifies a bucket, so it is not a wildcard permission.",
            },
            Object {
              "id": "AwsSolutions-KDF1",
              "reason": "The data stream is encrypted.",
            },
          ],
        },
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W86",
              "reason": "Retention period for CloudWatchLogs LogGroups are set to 'Never Expire' to preserve customer data indefinitely",
            },
            Object {
              "id": "W84",
              "reason": "By default CloudWatchLogs LogGroups data is encrypted using the CloudWatch server-side encryption keys (AWS Managed Keys)",
            },
          ],
        },
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "TestKinesisDataStreamKinesisFirehoseToS3firehoseloggroupfirehoselogstream1435D54F": Object {
      "DeletionPolicy": "Retain",
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-IAM5",
              "reason": "It specifies a bucket, so it is not a wildcard permission.",
            },
            Object {
              "id": "AwsSolutions-KDF1",
              "reason": "The data stream is encrypted.",
            },
          ],
        },
      },
      "Properties": Object {
        "LogGroupName": Object {
          "Ref": "TestKinesisDataStreamKinesisFirehoseToS3firehoseloggroup044C9C02",
        },
      },
      "Type": "AWS::Logs::LogStream",
      "UpdateReplacePolicy": "Retain",
    },
    "TestKinesisDataStreamKinesisStream0CE70927": Object {
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-IAM5",
              "reason": "It specifies a bucket, so it is not a wildcard permission.",
            },
            Object {
              "id": "AwsSolutions-KDF1",
              "reason": "The data stream is encrypted.",
            },
          ],
        },
      },
      "Properties": Object {
        "RetentionPeriodHours": 24,
        "ShardCount": 1,
        "StreamEncryption": Object {
          "EncryptionType": "KMS",
          "KeyId": "alias/aws/kinesis",
        },
        "StreamModeDetails": Object {
          "StreamMode": "PROVISIONED",
        },
      },
      "Type": "AWS::Kinesis::Stream",
    },
    "TestKinesisDataStreamKinesisStreamGetRecordsIteratorAgeAlarm7A8DFE52": Object {
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-IAM5",
              "reason": "It specifies a bucket, so it is not a wildcard permission.",
            },
            Object {
              "id": "AwsSolutions-KDF1",
              "reason": "The data stream is encrypted.",
            },
          ],
        },
      },
      "Properties": Object {
        "AlarmDescription": "Consumer Record Processing Falling Behind, there is risk for data loss due to record expiration.",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "EvaluationPeriods": 1,
        "MetricName": "GetRecords.IteratorAgeMilliseconds",
        "Namespace": "AWS/Kinesis",
        "Period": 300,
        "Statistic": "Maximum",
        "Threshold": 2592000,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "TestKinesisDataStreamKinesisStreamReadProvisionedThroughputExceededAlarm466027EF": Object {
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-IAM5",
              "reason": "It specifies a bucket, so it is not a wildcard permission.",
            },
            Object {
              "id": "AwsSolutions-KDF1",
              "reason": "The data stream is encrypted.",
            },
          ],
        },
      },
      "Properties": Object {
        "AlarmDescription": "Consumer Application is Reading at a Slower Rate Than Expected.",
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 1,
        "MetricName": "ReadProvisionedThroughputExceeded",
        "Namespace": "AWS/Kinesis",
        "Period": 300,
        "Statistic": "Average",
        "Threshold": 0,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "TestKinesisKinesisStreamsRoleDDF9DCCB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "firehose.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "kinesis:DescribeStream",
                    "kinesis:GetShardIterator",
                    "kinesis:GetRecords",
                    "kinesis:ListShards",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::GetAtt": Array [
                      "TestKinesisDataStreamKinesisStream0CE70927",
                      "Arn",
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "KinesisStreamsRoleRolePolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestLoggingBucket": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AccessControl": "LogDeliveryWrite",
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
